src/Cake.Common/IO/FileCopier.cs

```diff
@@ -103,10 +105,28 @@ public static void CopyFiles(ICakeContext context, IEnumerable<FilePath> filePat
                 throw new DirectoryNotFoundException(message);
             }
 
-            // Iterate all files and copy them.
-            foreach (var filePath in filePaths)
+            if (preserverFolderStructure)
             {
-                CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath));
+                var matches = from len in Enumerable.Range(0, filePaths.Min(s => s.ToString().Length)).Reverse() ////http://stackoverflow.com/a/8578268/2845744
+                              let possibleMatch = filePaths.First().ToString().Substring(0, len)
+                              where filePaths.All(f => f.ToString().StartsWith(possibleMatch))
+                              select possibleMatch;
+
+                var commonPath = System.IO.Path.GetDirectoryName(matches.First());
```

Comment:
You should not use `System.IO` operations here. They assume a physical file system.